Improve PID Tuning tab style and layout

IvoFPV opened the pull request:
The layout of the PID Tuning tab was changed to make it more organized. There is a 'Rateprofile Settings' subtab which now separated PID Profile and Rateprofile settings. Overall looks have been made more consistent, like table coloring, borders, separation, etc.

UPDATE: Sliders have been removed from the PR.

IvoFPV commented:
![Capture](https://user-images.githubusercontent.com/50072760/63209855-8ea99400-c0e6-11e9-8c8a-2d5a5ee1c27c.PNG)
![Capture2](https://user-images.githubusercontent.com/50072760/63209856-8ea99400-c0e6-11e9-8425-641dd99db072.PNG)
![Capture3](https://user-images.githubusercontent.com/50072760/63209857-8f422a80-c0e6-11e9-89f1-24953a4d9a0b.PNG)

McGiverGim commented:
It seems very promising! I will look into it deeply this week.

mikeller commented:
This does look promising indeed, thank you very much.
I think one thing that will be important for this will be to have it tested thoroughly for as many hardware combinations as possible. For this reason, my proposal is to, instead of putting it in relatively late for 4.1, hold this change back until 4.1 and the accompanying configurator have been released, and then put it in nice and early for 4.2, so that we have time to get feedback from users of the development configurator, and incorporate changes based on this feedback.
Thoughts?
(This will also enable us to stick to the common practice of adding MSP changes for new features only one version after the feature has been added, to reduce the likelihood of the MSP format needing changes because the features are not stable yet.)

IvoFPV commented:
I think the current version could work just fine without MSP support for defaults which are required for this to work. They are hardcoded for now but that should be updated for the next version.

Testing should be done on some things like F to D ratio and maximum P/D gain, but filtering sliders, P/D ratio slider don't really require much testing. Since everything is derived from the defaults which are definitely tried and confirmed to work great we just need to define maximum slider values which should stay within a safe limit. This is meant to be used by beginners who just want to have an easy way to tune their PIDs and filters. It shouldn't allow settings to be pushed all the way which is the main concern here. By keeping the values under the safe limit we remove any possibility of having bad things happen, like flyaways. So by defining FD ratio and most importantly maximum allowable D gains these sliders should fit most users needs. Those who have specific quads that the slider based values are not adequate possibly like micros or 7inch quads can just use the normal way of tuning instead. 

If the sliders are too big of a feature to be tried and tested before release of 4.1 maybe we could split the PR up in two parts, so only a new layout gets in for 4.1 and the sliders stay for 4.2.

mikeller commented:
@IvoFPV: The concerns I have with the sliders are to make sure they are ready to be used by beginners - they should not result in unexpected / dangerous behaviour if they are used with any non-exotic build. And this is what I think should get some testing.
So yes, splitting the redesign and the sliders into two pull request is definitely the better approach - even if we then decide that the sliders are safe to be put into 4.1.

IvoFPV commented:
@mikeller Should I just go ahead and split it right now or should we wait for @McGiverGim to give his thoughts on this?

mikeller commented:
@IvoFPV: Wait for @McGiverGim, he wanted to look at this.

McGiverGim commented:
This weekend I'm out of the computer. I will try to give my comments tomorrow Monday.

mikeller commented:
@IvoFPV: The code looks good to me, but I think the sizing of the 'Rates Preview' in the 'Rateprofile Settings' tab isn't working for non-large screen size (this is with a 1200 x 980 window):

![image](https://user-images.githubusercontent.com/4742747/63264619-a4ce7480-c2df-11e9-83b9-9df6808f4c6d.png)

It might be better to push the 3D model preview down below the other settings, or at least set a minimum width for the Rates Preview.

IvoFPV commented:
@mikeller Something like this?
![image](https://user-images.githubusercontent.com/50072760/63269370-5d6fc600-c296-11e9-97ef-ca156f1f706a.png)

McGiverGim commented:
I was looking at it and I noticed the same than @mikeller.

First of all, good job. I like the idea behind all. And second, this is only my opinion, I don't know if others will think the same. Better to wait to other opinions before making changes.

I understand that you have tried to include all the info without scrolls, but I think this is a lot of information in one visual and sometimes without order. Sometimes is better to include a scroll and let the user navigate in a predictable way. I think is more visual and more relaxed for the user and this will let grow some fields in the future without a complete refactor. 

PIDs tab:
- I don't see the division left/right of the PID tab. I understand that you were trying to include all in a page without scroll, but this has ended dividing similar elements at different places. For example, the antigravity is a PID element that now it is not with the other PID elements like ITerm Relax.
- I think is better to let all of them in only one column and maybe integrate the antigravity and dmin values in the same place than the other PID elements. In this way you will fix too the line break of the ITerm Relax and other elements at the min width (maybe some produced by translations).
![image](https://user-images.githubusercontent.com/2673520/63268694-04ebf900-c295-11e9-9f9c-fe2ca66378c9.png)
- Maybe you can, for the D-Min do in the same way than the ITerm Relax: add a check to enable/disable it, with the gain and advance like inputs inside. This check will show or hide the d-min column of the PID tab too.
- When I change the P values for example, the slides does not move with them, why? I suppose that they can reflect the new ratio.
- P/D ratio help: is there a consensus that a greater P/D ratio helps with propwash? I'm not too sure. A greater P/D gain yes.
- Is strange to me that the response slider changes the D value too, I expected only the FF but I suppose it has been discussed at Slack.

Rates tab:
- Since my point of view, three columns maybe is too much. Now when I open the tab I see too much information in too few space and it has the commented problem at min width. I think your latest capture is ok.
- To me the idea is like the old tab, inputs to the left, graphs at the right, but if the size of the elements don't let them be aesthetically correct, I accept the inputs at the top and the graph at the bottom without problem. 
- I don't see the TPA in the RATES tab. To me the TPA has more to do with the PID than with the RATE.

Some graphical glitches:
- The headers and other background colors don't fill the entire row.
![image](https://user-images.githubusercontent.com/2673520/63269424-79736780-c296-11e9-84f8-d90adda922b8.png)
- Some elements have border and rounded, others don't. Better if all the same.
![image](https://user-images.githubusercontent.com/2673520/63269615-d96a0e00-c296-11e9-9a00-8894dedb5d9c.png)
- Maybe we need two help icons, depending on the background. At some of them is difficult to see:
![image](https://user-images.githubusercontent.com/2673520/63269828-2e0d8900-c297-11e9-8d38-83b7ff1552bd.png)
- To me the circle of the sliders is too big. Maybe a little smaller or give them some more up and down margin.

Some of the problems commented here have been inherited from the old tab, but if we are refactoring it, I think is the moment to fix all if possible.

About the sliders, I like the idea, maybe in a future we can have only the sliders if they cover the major part of tuning but as suggested by @mikeller maybe is better to simply do a refactor of the elements here, merge it without problem, and then add the sliders to discuss the values, etc.

A lot of comments, but again, a great work @IvoFPV I have fighted several times against this tab and I know is a little nightmare.

IvoFPV commented:
@McGiverGim 
Yes, I was going for a more filled look without a need to scroll. Like the PID table doesn't need to span the whole width, just a third would be enough.

I was planning on integrating dMin and anti gravity in the PID controller settings box. That's a good point.
Dmin could be a switch like iTerm relax which hides the dmin from the PID table.

Problem with manually changing values is that they can't be reflected to slider positions correctly. If you only change P/D ratio on Roll the you have two different ratios. Currently they average value is put on the slider, only on refresh though, could be reworked to update on input easily.

P/D ratio help and all other help messages should be reviewed again and updated.
Response slider boosts FF and to compensate for FF induced overshoot it also scales Dmax values per FDratio. This has been discussed and found to work great.

Rates tab:
Yeah, it is updated now and I think it's not as compact but it provides better usability for sure.
TPA is a PID thing definitely and that should be changed in Firmware. Currently everything that is controlled by a Rateprofile is in the rates tab.

I will work on this some more to fix all those little rough edges you mentioned like borders and all other issues above. 
Sliders in their current state have been working great and the values were discussed on slack with @ctzsnooze, @spatzengr and others. They provide just enough tuning range without putting beginners in danger of too little filtering or high D gains. With good tooltip messages they should be super easy to use.

However, I agree we should split the refactor and the sliders into two PRs. I will just continue working on the refactor on this branch and when you agree that it is acceptable for a merge I will just remove the sliders which is really easy to do and open a second PR.

Thanks for a lot of input, this is what I needed the most.

McGiverGim commented:
> Problem with manually changing values is that they can't be reflected to slider positions correctly. If you only change P/D ratio on Roll the you have two different ratios. Currently they average value is put on the slider, only on refresh though, could be reworked to update on input easily.

This is one of the reasons why maybe we need to split it in two PR, to discuss thinks like that. Maybe we need another slider to mark the difference between roll and pitch, for example. It is not the same the difference between a racer or a freestyle quad. But better discuss it later to finish this first ;)

> TPA is a PID thing definitely and that should be changed in Firmware. Currently everything that is controlled by a Rateprofile is in the rates tab.

I didn't know that :( if it is controlled by the rateprofile must be in the rates tab until changed in firmware as you said.

IvoFPV commented:
@McGiverGim I moved dmin and anti gravity in the PID controller settings box. I'm not sure on this full width design. I would prefer to have two columns, left side PID values box and sliders, right side, PID controller settings box and Angle, MAG, GPS stuff

McGiverGim commented:
Yes, I have tested your code, full width is too much. Other tabs have a "max-width" defined to let some empty space to the right, for example a max-width of 950px or similar, adding this it looks better. But we can explore other solutions. 

I think that your suggestion is interesting, if we can put all the data in two columns in a reasonable way when we made the Configurator at the minimum size, is a great idea. If you get to something that looks good, is the best solution. If you have time to try, if I have understood, is:
- Left: PIDs + Sliders
- Right: PID Controller settings + PID angle, mac, etc.

I think I would put the PID angle in the left too, because the right part will grow in time with new features, but is better to try and see how it looks.

The only problem with this is that maybe we need to "modify" all the PID Controller settings elements to fill in the right with Configurator window at reduced size.

Try it, if it does not fit, maybe we can change all of them to be the same (this will take more work on the JS part, so only try to do it if you see that the other way does not fit and you have time):
- Checkbox to enable, like ITerm Rotation.
- If it has parameters, in a second row, under the label, add one input and slider by parameter (or only an input). The good about this is that if can be reduced without too much problem and will look good.

It will take more vertical space (each parameter will be one row) but less horizontal and they will look more "all the same".

IvoFPV commented:
I really don't like the overall look of the PID controller settings table. Sliders are kinda redundant there and take a lot of horizontal space, causing issues with smaller windows sizes, especially with long translations like you pointed out. Maybe I should just do a complete refactor of that table and try to make it look good. It won't be that much work I hope.
This is what I think you meant:
--------------------------
Iterm relax -> Switch
--------------------------
axis. ___ type: ___ cutoff:__

So labels like iterm relax, dmin are shown with its switch to enable it and when you enable it below the label its settings are then shown. Otherwise they are not.

McGiverGim commented:
Yes, what is I wanted to say with the checkbox, they hide the settings (and assign a value of zero usually).
But I liked the sliders, but maybe you have reason and is better to remove them. Tomorrow I will try to draw something but now I'm out of the computer.

IvoFPV commented:
Looks good to me, unnecessary sliders removed, fixed the resizing issue.
![image](https://user-images.githubusercontent.com/50072760/63307547-0d711d80-c2ef-11e9-9e50-dbb411a6888a.png)

McGiverGim commented:
Yes, I think we are near... The new disposition has sense and it adjust more or less correctly to a little window.
I don't like the new style, it has a lot of horizontal and vertical lines between elements and this makes it less elegant than the current version, but maybe we need to add them now that we have the elements in the second row.

My initial idea was something like:
![image](https://user-images.githubusercontent.com/2673520/63330763-ae35fc00-c334-11e9-9a7c-cc210a9c0a31.png)

But maybe I'm the only one who likes the sliders and is better to use only inputs as you have made. The problem with that is that any element has a different size. We need to make them all similar.

These are only my suggestions, I will try:
- Adjust the checkbox/switch to the top, to be always aligned with the title of the feature, not like now that it is at the middle.
- The antigravity switch is not working. I suppose that is because it is enabled in the Features, but we can accomplish the same if when the switch is disabled, we assign a gain of 1000 to it.
- I will try to align the name/input of the parameters. The easiest way to do it is to let the input/select at the left and the text at the right, and add a width and max-width to the inputs (in this way the selects and inputs will have the same size). With a little indentation like you have now.

Something like this:
![image](https://user-images.githubusercontent.com/2673520/63331498-1a652f80-c336-11e9-8a8a-1b6a41988470.png)

I think in this way maybe we can remove some of the lines (maybe the vertical, maybe the horizontal, or maybe neither of them).

IvoFPV commented:
@McGiverGim Antigravity is disabled as input since I didn't have time to try to implement feature disabling and enabling in the tab. This would require restart and so on. Better to make the switch set it to 1000 which is basically disabled. I will do that.
I will place the inputs on the left to align the suboptions, but the problem is the selects which are sized per its longest option label. We can force the lower width but then the text would be cut out, like RP (increment only) is a long text and won't fit if we make it narrower.
I like the lines but maybe they could be less intrusive, so I will make them less visible by using a darker color.
Regarding the horizontal space, it is wide right now and not the best but I am thinking that in future we can use two tables side by side. Essentially this will cut the width in half when we start adding other features. So it would be half for the PID values, sliders, and two fourths for pid controller settings.

McGiverGim commented:
About the selects, I have done it in other places of the Configurator. As you said, the option is cut, but when you press on it to view all the options they appear with the complete text. I prefer this and make them aligned.
If not, at least aligning it on the left will end with two different sizes: one for inputs and other for selects, but I think it will be more consistent.

IvoFPV commented:
Here I fixed the issue with alignment of the inputs and selects. Label has a min width. If there are longer translations it will just be unaligned again, but for most it will be aligned correctly. What do you think?

McGiverGim commented:
It depends a lot of the translations. There are languages that need longer texts and others shorters one. And for some reason, at minimum size they are totally misaligned having a lot of space to the right. For example the d_min:

![image](https://user-images.githubusercontent.com/2673520/63339519-5c49a200-c345-11e9-91a1-ba56d584aa6e.png)

The antigravity is misaligned too, I know you have modified the text to be shorter, but some languages may need more text than others, so it is always let space if we can.

IvoFPV commented:
@McGiverGim Then maybe inputs and selects on the left like you said. This way we assure more space for the label. It will never be perfect. Some translations are super long and will need to wrap.
Thats why I said that it will look great on most translations where labels are short, but unalignment in some small number of cases is okay to me...
![image](https://user-images.githubusercontent.com/50072760/63340268-ecd4b200-c346-11e9-8347-2e6e0ff3dd8c.png)

McGiverGim commented:
To me is ok (we will need to remove the `:` in the texts but for the rest is aligned) but I don't know what others think about this. Can you post the entire capture to see the whole picture and let others give his opinion?

IvoFPV commented:
![image](https://user-images.githubusercontent.com/50072760/63341249-168ed880-c349-11e9-8958-cc1a77485778.png)
This looks good to me. But again there will be issues with text being cutout for longer select item translations.

McGiverGim commented:
Is good to me now. And yes, if someone makes a very long translation, it will look strange, but now we have a lot of room space before this happens.

Some suggestions to look better, they are only my thoughts so no need to implement all of them: 
- Give a little of border-round to the inputs and selects, like it is being done in other places of the Configurator.
- Start the gain in the anti gravity in 1.1. If someone wants a value of 1 he needs to disable the switch.
- Maybe in a future change all of the features (absolute control, acro trainer, etc. ) to the same style of ckeckbox + input
- Maybe limit the width of the tab to a something reasonable, letting empty space on the right. It will look far better.
- To me is strange the filters elements, with a 50% for left (switch and input) and 50% for right (texts). It will look better if the left part is shorter or maybe fixed size. Now the texts are being cut when there is enough space for them.

For the rest I like it. Good work.

IvoFPV commented:
Added the borders on inputs and selects and rounded them. AG is now fixed. Rates tab has been improved. Previous was too wide in my opinion. Now it wraps the rate preview model when resizing.
I think this now looks pretty solid.

McGiverGim commented:
It is going better version by version :)

When I was talking about limiting the tab width, I was not talking only about the rates tab. I think all of them improve it they are not too wide, but is only my opinion.

And when I was talking about the filters tab, I was talking about this:

![image](https://user-images.githubusercontent.com/2673520/63351655-eacb1d00-c35f-11e9-9edf-94e623750ea1.png)

This was like this before your PR, so is not your fault, but having the 1 and 2 parts of the same size, makes it look a little strange and cuts long strings again in two when we have A LOT of space to the left. Simply adjusting it with a fixed size of maybe a better percentage (35%-65% for example) and adding a min-width to be sure not to break the slider+input field if too small will be enough. What do you think?

I've observed too that the help messages for d_min and antigravity are missing.

IvoFPV commented:
@McGiverGim Yeah, I know you weren't talking only about the rates tab but I had to fix that first because it was really bad in my opinion. PID profile tab looks a bit wide too because of the new table now which has a lot of horizontal space. Don't know how should I fix that. Just reduce the whole width of that tab and center it or align it left?
I also don't like filter table columns 50/50 width and will look into that next.
I am aware that the help messages are missing, but that will be the last part to work on.

McGiverGim commented:
> PID profile tab looks a bit wide too because of the new table now which has a lot of horizontal space. Don't know how should I fix that. Just reduce the whole width of that tab and center it or align it left?

If you want to fix it you must try and error. I think is better maybe to let the left table (PIDs) be a little wider than the right table (maybe 60-40) but the right table must be aligned to the left. Centered will be like detached and we don't want that. But this is only adjust to finest, like it is now is good enough. I will try the 60%-40% thing and if later you reduce the total tab width (think that some people for example have 27" or more with some super wide version) I think it will be superb, but without trying I can't say for sure. Now that it is good is a little what you like too.

spatzengr commented:
For consideration in column/slider naming.  I know the tooltips have some documentation on the P,I, and D gains (which needs some help too) but was hoping we could add some short layman's terms into the headers directly as I'm sure not everyone notices and goes to the tooltips right away.  My thought is the direct labeling should be more layman's terms and the tooltips can give more depth and specifics.  More related to flight characteristics changes of each item and less related to the coding behind it.

Do note, the filters markup is 1 version old, so now we just have one slider.  So we only need better naming for the Low and High filter columns above the slider IMHO (more descriptive).  Trying to get the point across on the trade-offs with filtering to someone brand new to the BF Configurator and kwad flying.

Ideally, we can express enough through good column naming (flight characteristics based) and tooltips that folks can get a gist of how adjustments will change their flight characteristic and don't need to decode/translate desired flight characteristics to Configurator knobs and pullies through a JB or UAV Tech video.

![BF Configurator Markup_1](https://user-images.githubusercontent.com/25570978/63354866-31eee900-c333-11e9-97c4-9167f8c477cf.png)
![BF Configurator Markup_2](https://user-images.githubusercontent.com/25570978/63354874-35827000-c333-11e9-9cec-e6d67f57c9fd.png)

McGiverGim commented:
@spatzengr you are working over a very old version of the PR, but I think you are right at some questions. For example the naming of the new PID sliders. But I have not entered there because we have decided to remove them from this PR, merge it with the new style and ordering and then @IvoFPV will open a new PR to discuss about the sliders, values, defaults, etc.

spatzengr commented:
Yes, photos were old.  I sent to @IvoFPV  to be included in the PR, but he wanted me to post on the PR for others review and comment.  I will refresh on the latest version, wait for the new PR and re-post.

IvoFPV commented:
@McGiverGim Something like this? Or maybe not so much to the left?
![image](https://user-images.githubusercontent.com/50072760/63358595-9cbc1680-c36b-11e9-8b80-70988ffe0717.png)

McGiverGim commented:
To me looks good. Of you use a lower resolution (or finally you decide to limit the width of the tab) you will see that the amount of free space is lower than your capture, si ir looks ok

IvoFPV commented:
@McGiverGim So is everything good now? Just one more thing that bothers me is this. 
![image](https://user-images.githubusercontent.com/50072760/63368378-10671f00-c37e-11e9-90b8-3d5fedf2dae1.png)
You pointed that out but I've tried to track it down and fix it but wasn't able to. Seems to be some border-collapse thing I just can't solve. Maybe you could take a look into that part.

IvoFPV commented:
I've fixed the table border gap. Also, iterm relax type was being reset when disabled to Gyro, this was useful on previous versions because that was the default setting, now in 4.0+ when you disabled iterm relax and then enabled it it would be set to Gyro, even though default is Setpoint. It is better to have it just not change that on disabling. Makes it more consistent.

I can't find any more cosmetic issues. This should now be ready.

McGiverGim commented on locales/en/messages.json:
```diff
@@ -1456,40 +1462,29 @@
+       "description": "Advance of the D Min feature"
+    },
+    "pidTuningDMinHelp": {
+        "message": "Gain adjusts how fast D min gets up to its maximum value and is based on gyro to determine sharp moves and propwash events. Advance makes D min go up earlier by using setpoint instead of gyro to determine sharp moves."
```
This is a general comment, not attached to any particular message. We are trying to give context to all the messages for the translations. Maybe some of them not need it, but if if in doubt, is a good habit to add a description with the context , the element and or the explanation of the message.

IvoFPV replied:
I will add descriptions for messages where actual messages lack context.

IvoFPV commented on src/css/tabs-dark/pid_tuning-dark.css:
```diff
@@ -19,7 +19,7 @@
 
 .tab-pid_tuning .cf tr {
-    background-color: #393b3a;
+    background-color: #414443;
```
I will go through colors and see where it can be replaced with variables.

IvoFPV replied:
There are a lot of places where variables can be used but I am not really into changing all that now. Maybe that could be done another time.

IvoFPV commented:
@McGiverGim DMin setting to 0 when you enable it is what I don't like also. Maybe we could do current Dmax - 1 as a value when you enable it, to me this is kinda clunky, but let's see what others think.

McGiverGim commented on locales/en/messages.json:
```diff
@@ -1444,8 +1447,12 @@
     "pidTuningAntiGravityMode": {
-        "message": "Anti Gravity Mode"
+        "message": "Mode",
+        "description": "One of the modes of anti gravity"
```
I think this is more one of the "parameters" of antigravity, not one mode.

McGiverGim commented on locales/en/messages.json:
```diff
@@ -1456,40 +1463,32 @@
+       "description": "Advance of the D Min feature"
+    },
+    "pidTuningDMinHelp": {
+        "message": "Gain adjusts how fast D min gets up to its maximum value and is based on gyro to determine sharp moves and propwash events. Advance makes D min go up earlier by using setpoint instead of gyro to determine sharp moves.",
```
Maybe the D Min must explain a little, in one phrase, what is D Min and after that, comment the two parameters as you have did.

IvoFPV replied:
Added the short description.

IvoFPV commented:
For D min, there are several options for its value when it gets enabled in my opinion.
1. Dmin gets set to 0 like it is now
2. Dmin gets set to Dmax - 1 value
3. Dmin get set at a specific ratio, like half of the Dmax
Defaults are not going to work because someone could have Dmax lower than defaults.
Also someone may use Dmin on specific axes. Like only Roll, or even all of them, including yaw.

McGiverGim commented:
True, I didn't thought that can be enabled only by axes... Then to me is right to let it as zero. Maybe the switch has not sense then ☹️
Let's see other opinions.

IvoFPV commented:
I did some cleaning up of CSS. Dmin is the only thing left to figure out now. @mikeller, any thoughts?

McGiverGim commented:
> However, on small window sizes, the PID table up top is not visible together with the D Min switch, making it hard for the user to find out that they are expected to set up the per axis D Min values after enabling D Min - I think we should add a notice about this to the D Min box:

I think this is new. In earlier version the box where always left-right, always inside the screen. Now it is adaptive.

IvoFPV commented:
I've added the d min disabled note above the PID table. Fixed integrated yaw space. Improved wrapping so the spacer gets removed.
Also, why are your select boxes looking gray? @mikeller 
For me they are white like they should be.
![image](https://user-images.githubusercontent.com/50072760/63633982-edc05900-c650-11e9-9199-55654ad92a6c.png)

ctzsnooze commented:
Several users have got messed up unintentionally with the current configurator setting D min to zero if they play with low D numbers.  
I would be very happy if D min moved in tandem with D itself.  
It could follow the default percentage (whatever that is).  
My experience has been that for racing, D min of about half D is good; for freestyle, D min at about ⅔-¾ of D works better.  I think the default is a sort of compromise around ⅔.

bizmar commented:
Finally had time to look it over after some time.

### Pids tab:
D_min and D are stil not presented well. It's a hard problem. This idea is not the best, I'm sure a there is a even better way of denoting that Derivative now has two values to adjust. (quick photoshop butchery)
![Pidstab](https://user-images.githubusercontent.com/4053708/63637752-ca150700-c680-11e9-8f30-244d941d1335.jpg) Maybe add a switch to Pid controller settings called "enable yaw Derivative", now we have 2x the opportunity to miss a field and have an instant flyaway. Doing that complicates enabling Integrated yaw... Idk. (imho the value of integrated yaw is still an open topic, and should not have been put into the UI in the first place)

Is the order of pid controller settings important?
Maybe bring D_min further up, so there's less chance of it being cut off. (yes, the list almost looks nice now that all the options are grouped by input type). Maybe call it "Dynamic Derivative" or something else, if the name in the PID table is changed, D_min makes a bit less sense. D_min is a very CLI sounding setting... I'm sure there's a middle ground between "D_min" and "absolute control" where a feature can have a good name and be intuitively understandable.

Feedforward 2.0 adds some settings that should be in the UI, and may merit it having it's own sub-section like D_min or item relax.

In general, if there is any will to reorder the list, I would push the more obscure features down, and the often changed ones up. And make the whole integrated yaw "click here" help sentence a link.

Some parts of the configurator that are "conceptually" connected are very "far apart" like having to enable ACC, before you can even see auto-level settings, maybe that section can be collapsed and grayed out with "enable accelerometer" or similar written on top. Lookin forward, RPM based features will probably have the same issue. RPM telemetry is exactly like Accelerometer, it's a "info source" that enables features, but has to be turned on for the features to work.


### Rates tab:
It feels weird, but all change does. So let's go with it. I mostly run my BF conf. full screen on my 15" laptop, that reflows it into a 3 column layout that is a bit strange. The two column one is ok. No real opinions on what to change, it's just unfamiliar. 


### Filers tab:
It would be great if disabled fields had the same style. Like this one:
<img width="576" alt="Screenshot 2019-08-24 at 14 06 23" src="https://user-images.githubusercontent.com/4053708/63638277-5d047000-c686-11e9-8601-df2b3d88ab9d.png">
If dyn filter had a switch that would be nice too. Are there "initialization" issues with that? Are they significant?

### Other:

When I first open the PID tab, some sections are misaligned, then they reflow when I adjust window size.
<img width="350" alt="Screenshot 2019-08-24 at 16 14 04" src="https://user-images.githubusercontent.com/4053708/63638565-4bbd6280-c68a-11e9-9623-f52ede0f72b9.png">


The 45º caution tape stripes... "Watch out! There's some serious settings here!"
I like them, but they are already a hassle to deal with.

If we look at these two 10.5 screenshots, there are 4 different heading and "stripe" combinations.. 
<img width="642" alt="Screenshot 2019-08-24 at 16 57 28" src="https://user-images.githubusercontent.com/4053708/63639230-27b14f80-c691-11e9-9ad2-71b316e80be7.png">
<img width="473" alt="Screenshot 2019-08-24 at 16 57 44" src="https://user-images.githubusercontent.com/4053708/63639232-2b44d680-c691-11e9-8ef5-755d048e26f9.png">
Getting rid of the stripe and using dark gray as "H1" and lighter gray as "H2" seems the most consistent. Maybe the basic/acro and angle/horizon sections can be special... The title should be on top in any case.

bizmar commented:
@ctzsnooze this PR now "sets the table" for sliders. And has to function on its own if sliders will be pushed to 4.2.

IvoFPV commented:
@bizmar Thanks for the input.

> D_min and D are stil not presented well. It's a hard problem. This idea is not the best, I'm sure a there is a even better way of denoting that Derivative now has two values to adjust. (quick photoshop butchery)

This looks much better and should be much more intuitive.

> D_min is a very CLI sounding setting... 

D min should be renamed in my opinion to something like you said, Dynamic Derivative

> Maybe add a switch to Pid controller settings called "enable yaw Derivative", now we have 2x the opportunity to miss a field and have an instant flyaway.

I don't think we need to do any restrictions like this, but if someone is changing any settings without knowing what they are doing then they shouldn't be doing them. Maybe we could add a note that says: 'It is not recommended to run D term on yaw and this feature is mainly focused for tricopters and may cause flyaways if used on a normal quad' or something like that.

> Feedforward 2.0 adds some settings that should be in the UI, and may merit it having it's own sub-section like D_min or item relax.

FF2.0 shoudn't be added to the configurator if we are following the 'new features get into GUI one version later' paradigm.

> If dyn filter had a switch that would be nice too. Are there "initialization" issues with that? Are they significant?

This could be done, but would require restart on saving. Also splitting the features on the configuration tab to pid tab isn't good in my opinion. Maybe it could be fixed on the firmware side so the dynamic notch can be enabled/disabled like any other filter.

> When I first open the PID tab, some sections are misaligned, then they reflow when I adjust window size.
 
This should now be fixed.

bizmar commented:
> I don't think we need to do any restrictions like this, but if someone is changing any settings without knowing what they are doing then they shouldn't be doing them. Maybe we could add a note that says: 'It is not recommended to run D term on yaw and this feature is mainly focused for tricopters and may cause flyaways if used on a normal quad' or something like that.

Maybe I overestimate the the possibility of this happening, only know myself.. And I know that I missed a field and put FF into D by accident. Twice. I don't really a have strong opinion about it. If I put my FF into D on Roll or Pitch it would probably flyaway too.

> Feedforward 2.0 adds some settings that should be in the UI, and may merit it having it's own sub-section like D_min or item relax.

Yes, although, it's such a good improvement with no downsides, will probably be enabled by default, has been extensively flown and improved since April. I think it really deserves to be in the UI

I can point to Absolute control, Integrated yaw and i-term relax, by that rule, those were all added "too soon".  For i-term relax it world be a real shame if most if the community didn't have it for 4-6 months, and it was a real shame that a important setting was not exposed for that time. Thanks Ivo for adding it now. 

Anyway... I'm not here to blame anyone, I would just like to point out the rule was more of a guideline in some instances, and that I think FF2.0 deserves an exemption, as does Dynamic Idle.

mikeller commented:
@IvoFPV: The drop-downs are always grey for me - I guess that's just the styling that is used in linux. I've never read much into it, and it's a different grey than what is used when a drop-down is disabled.

Not sure what we want to do here. I'd be happy to merge this as it is, and leave more improvements for future pull requests. But we can also work some more on this one, and try to make it more user friendly. @IvoFPV, @McGiverGim, @bizmar, @ctzsnooze thoughts?

IvoFPV commented:
@mikeller I think the functionality and layout part is complete, so it can be merged. Other things like naming or positioning of Dmin differently, improving of messages or any other small change like that could be done in another PR. Also I am going to PR the sliders feature after this has been merged, so further progress could be done there as a part of making it more user friendly.
